// @xliic/common/jsonpointer.js
"use strict";
/*
 Copyright (c) 42Crunch Ltd. All rights reserved.
 Licensed under the GNU Affero General Public License version 3. See LICENSE.txt in the project root for license information.
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.deref = exports.find = exports.findByPath = exports.parseJsonPointer = void 0;
const SLASHES = /\//g;
const TILDES = /~/g;
function parseJsonPointer(pointer) {
    const hasExcape = /~/;
    const escapeMatcher = /~[01]/g;
    function escapeReplacer(m) {
        switch (m) {
            case "~1":
                return "/";
            case "~0":
                return "~";
        }
        throw new Error("Invalid tilde escape: " + m);
    }
    function untilde(str) {
        if (!hasExcape.test(str)) {
            return str;
        }
        return str.replace(escapeMatcher, escapeReplacer);
    }
    return pointer.split("/").slice(1).map(untilde).map(decodeURIComponent);
}
exports.parseJsonPointer = parseJsonPointer;
function findByPath(target, path) {
    let current = target;
    for (const segment of path) {
        if (typeof current === "object" && current !== null) {
            if (Array.isArray(current)) {
                const index = parseInt(segment, 10);
                if (isNaN(index)) {
                    return undefined;
                }
                current = current[index];
            }
            else if (current.hasOwnProperty(segment)) {
                current = current[segment];
            }
            else {
                return undefined;
            }
        }
        else {
            return undefined;
        }
    }
    return current;
}
exports.findByPath = findByPath;
function find(target, pointer) {
    if (Array.isArray(pointer)) {
        return findByPath(target, pointer);
    }
    return findByPath(target, parseJsonPointer(pointer));
}
exports.find = find;
function deref(oas, maybeRef) {
    if (maybeRef === undefined) {
        return undefined;
    }
    if ("$ref" in maybeRef) {
        const refTarget = find(oas, maybeRef.$ref);
        return refTarget;
    }
    return maybeRef;
}
exports.deref = deref;
